src/credentials/tests/TestGroupDataProvider.cpp

```diff
@@ -16,209 +16,937 @@
+{
+    chip::TestPersistentStorageDelegate delegate;
+
+    char out[128];
+    uint16_t size = static_cast<uint16_t>(sizeof(out));
+
+    NL_TEST_ASSERT(apSuite, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND == delegate.SyncGetKeyValue(kKey1, out, size));
+
+    size = static_cast<uint16_t>(kSize1);
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == delegate.SyncSetKeyValue(kKey1, kValue1, size));
+
+    size = static_cast<uint16_t>(sizeof(out));
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == delegate.SyncGetKeyValue(kKey1, out, size));
+    NL_TEST_ASSERT(apSuite, size == kSize1);
+    NL_TEST_ASSERT(apSuite, !memcmp(out, kValue1, kSize1));
+
+    size = static_cast<uint16_t>(kSize2);
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == delegate.SyncSetKeyValue(kKey1, kValue2, size));
+
+    size = static_cast<uint16_t>(sizeof(out));
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == delegate.SyncGetKeyValue(kKey1, out, size));
+    NL_TEST_ASSERT(apSuite, size == kSize2);
+    NL_TEST_ASSERT(apSuite, !memcmp(out, kValue2, kSize2));
+
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == delegate.SyncDeleteKeyValue(kKey1));
+
+    size = static_cast<uint16_t>(sizeof(out));
+    NL_TEST_ASSERT(apSuite, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND == delegate.SyncGetKeyValue(kKey1, out, size));
+}
+
+constexpr chip::FabricIndex kFabric1  = 1;
+constexpr chip::FabricIndex kFabric2  = 7;
+constexpr chip::EndpointId kEndpoint1 = 1;
+constexpr chip::EndpointId kEndpoint2 = 0xabcd;
+constexpr chip::EndpointId kEndpoint3 = 0xfffe;
+constexpr chip::GroupId kGroup1       = kMinFabricGroupId;
+constexpr chip::GroupId kGroup2       = 0x2222;
+constexpr chip::GroupId kGroup3       = kMaxFabricGroupId;
+
+static const GroupMapping endpoint1group1(kEndpoint1, kGroup1, "Group 1.1");
+static const GroupMapping endpoint1group2(kEndpoint1, kGroup2, "Group 1.2");
+static const GroupMapping endpoint1group3(kEndpoint1, kGroup3, "Group 1.3");
+
+static const GroupMapping endpoint2group1(kEndpoint2, kGroup1, "Group 2.1");
+static const GroupMapping endpoint2group2(kEndpoint2, kGroup2, "Group 2.2");
+static const GroupMapping endpoint2group3(kEndpoint2, kGroup3, "Group 2.3");
+
+static const GroupMapping endpoint3group1(kEndpoint3, kGroup1, "Group 3.1");
+static const GroupMapping endpoint3group2(kEndpoint3, kGroup2, "Group 3.2");
+static const GroupMapping endpoint3group3(kEndpoint3, kGroup3, "Group 3.3");
+
+void TestGroupMappings(nlTestSuite * apSuite, void * apContext)
+{
+    GroupDataProvider * groups = GetGroupDataProvider();
+    NL_TEST_ASSERT(apSuite, groups);
+
+    // Reset test
+
+    groups->RemoveAllGroupMappings(kFabric1, kEndpoint1);
+    groups->RemoveAllGroupMappings(kFabric1, kEndpoint2);
+    groups->RemoveAllGroupMappings(kFabric1, kEndpoint3);
+    groups->RemoveAllGroupMappings(kFabric2, kEndpoint1);
+    groups->RemoveAllGroupMappings(kFabric2, kEndpoint2);
+    groups->RemoveAllGroupMappings(kFabric2, kEndpoint3);
```

Comment:
There should be a more scalable to do this, possilby with a test Setup method.

Reply:
The Setup function is only called before the first test function. In consequence we have two options:
1. Make sure that each function leaves the storage in a condition that is suitable for the next test, or
2. Each test function must ensure its own initial conditions.
I don't think a test function should rely on other test functions to work properly, so I stand by my choice of option 2, however, there is a simpler, more elegant way to reset any of these tests, which I am using now:
```
    groups->RemoveFabric(kFabric1);
    groups->RemoveFabric(kFabric2);
```